[BUG] Error connecting to secure OS cluster when using OSB on Python-3.10 and above.

**Describe the bug**
While trying to run OSB against a remote cluster with security enabled I am getting below error:
```
[ERROR] Cannot execute_test. Error in test execution orchestrator (ConnectionError(Cannot create a client socket with a PROTOCOL_TLS_SERVER context (_ssl.c:801)) caused by: SSLError(Cannot create a client socket with a PROTOCOL_TLS_SERVER context (_ssl.c:801)))
```
This error is only encountered on machines running Python-3.10 and above. 
I am able to execute OSB against secure remote cluster when running on machines with Python-3.9 and below. 

**To Reproduce**
Set up a machine with Python 3.10 or above and run OSB benchmarking on remote cluster using following command:
```
opensearch-benchmark execute_test --workload=geonames --test-mode --target-hosts=10.10.10.10:9200 --pipeline=benchmark-only --client-options="use_ssl:true,verify_certs:false,basic_auth_user:'admin',basic_auth_password:'admin'"
```

**Expected behavior**
The OSB benchmark should run without throwing any error. 

**Logs**
```
2023-04-03 20:22:11,457 -not-actor-/PID:74445 opensearch WARNING GET https://opens-publi-2nkfmfd5td5u-d3ceb505e4ca2889.elb.eu-west-1.amazonaws.com:443/ [status:N/A request:0.142s]
Traceback (most recent call last):
  File "/opt/homebrew/lib/python3.10/site-packages/urllib3/connectionpool.py", line 703, in urlopen
    httplib_response = self._make_request(
  File "/opt/homebrew/lib/python3.10/site-packages/urllib3/connectionpool.py", line 386, in _make_request
    self._validate_conn(conn)
  File "/opt/homebrew/lib/python3.10/site-packages/urllib3/connectionpool.py", line 1042, in _validate_conn
    conn.connect()
  File "/opt/homebrew/lib/python3.10/site-packages/urllib3/connection.py", line 414, in connect
    self.sock = ssl_wrap_socket(
  File "/opt/homebrew/lib/python3.10/site-packages/urllib3/util/ssl_.py", line 449, in ssl_wrap_socket
    ssl_sock = _ssl_wrap_socket_impl(
  File "/opt/homebrew/lib/python3.10/site-packages/urllib3/util/ssl_.py", line 493, in _ssl_wrap_socket_impl
    return ssl_context.wrap_socket(sock, server_hostname=server_hostname)
  File "/opt/homebrew/Cellar/python@3.10/3.10.8/Frameworks/Python.framework/Versions/3.10/lib/python3.10/ssl.py", line 513, in wrap_socket
    return self.sslsocket_class._create(
  File "/opt/homebrew/Cellar/python@3.10/3.10.8/Frameworks/Python.framework/Versions/3.10/lib/python3.10/ssl.py", line 1062, in _create
    self._sslobj = self._context._wrap_socket(
ssl.SSLError: Cannot create a client socket with a PROTOCOL_TLS_SERVER context (_ssl.c:801)
```

**More Context (please complete the following information):**
- Workload(Share link for custom workloads)
- Service(E.g OpenSearch)
- Version (E.g. 1.0)

**Additional context**
Add any other context about the problem here.

--- a/osbenchmark/client.py
+++ b/osbenchmark/client.py
@@ -223,7 +223,9 @@ class OsClientFactory:
                 has_hostname = True
 
         if has_ip and has_hostname:
-            logger.warning("Although certificate verification is enabled, peer hostnames will not be matched since the host list is a mix of names and IP addresses")
+            logger.warning("Although certificate verification is enabled, "
+                "peer hostnames will not be matched since the host list is a mix "
+                "of names and IP addresses")
             return False
 
         return has_hostname

--- a/osbenchmark/client.py
+++ b/osbenchmark/client.py
@@ -220,7 +220,8 @@ class OsClientFactory:
                 has_hostname = True
 
         if has_ip and has_hostname:
-            raise exceptions.SystemSetupError("Could not verify certs since both IP addresses and hostnames were provided. Please ensure one or the other is used.")
+            raise exceptions.SystemSetupError("Could not verify certs since both IP addresses \
+                 and hostnames were provided. Please ensure one or the other is used.")
 
         return has_hostname
 

--- a/osbenchmark/client.py
+++ b/osbenchmark/client.py
@@ -220,8 +220,7 @@ class OsClientFactory:
                 has_hostname = True
 
         if has_ip and has_hostname:
-            raise exceptions.SystemSetupError("Could not verify certs since both IP addresses \
-                 and hostnames were provided. Please ensure one or the other is used.")
+            raise exceptions.SystemSetupError("Can't verify certs since both IP addresses and hostnames were found.")
 
         return has_hostname
 

--- a/osbenchmark/client.py
+++ b/osbenchmark/client.py
@@ -155,7 +155,9 @@ class OsClientFactory:
                 # advised. See: https://urllib3.readthedocs.io/en/latest/advanced-usage.html#ssl-warnings"
                 urllib3.disable_warnings()
             else:
-                # check_hostname should be False if any host is an IP address
+                # The peer's hostname can be matched if only a hostname is provided.
+                # In other words, hostname checking is disabled if an IP address is
+                # found in the host lists.
                 self.ssl_context.check_hostname = self._has_only_hostnames(hosts)
                 self.ssl_context.verify_mode=ssl.CERT_REQUIRED
                 self.logger.info("SSL certificate verification: on")
@@ -212,6 +214,7 @@ class OsClientFactory:
 
     @staticmethod
     def _has_only_hostnames(hosts):
+        logger = logging.getLogger(__name__)
         has_ip, has_hostname = False, False
         for host in hosts:
             if is_ipaddress(host["host"]):
@@ -220,7 +223,8 @@ class OsClientFactory:
                 has_hostname = True
 
         if has_ip and has_hostname:
-            raise exceptions.SystemSetupError("Can't verify certs since both IP addresses and hostnames were found.")
+            logger.warning("Although certificate verification is enabled, peer hostnames will not be matched since the host list is a mix of names and IP addresses")
+            return False
 
         return has_hostname
 

--- a/osbenchmark/client.py
+++ b/osbenchmark/client.py
@@ -28,6 +28,7 @@ import time
 
 import certifi
 import urllib3
+from urllib3.util.ssl_ import is_ipaddress
 
 from osbenchmark import exceptions, doc_link
 from osbenchmark.utils import console, convert
@@ -138,16 +139,14 @@ class OsClientFactory:
             self.logger.info("SSL support: on")
             self.client_options["scheme"] = "https"
 
-            # ssl.Purpose.CLIENT_AUTH allows presenting client certs and can only be enabled during instantiation
-            # but can be disabled via the verify_mode property later on.
-            self.ssl_context = ssl.create_default_context(ssl.Purpose.CLIENT_AUTH,
+            self.ssl_context = ssl.create_default_context(ssl.Purpose.SERVER_AUTH,
                                                           cafile=self.client_options.pop("ca_certs", certifi.where()))
 
             if not self.client_options.pop("verify_certs", True):
                 self.logger.info("SSL certificate verification: off")
                 # order matters to avoid ValueError: check_hostname needs a SSL context with either CERT_OPTIONAL or CERT_REQUIRED
-                self.ssl_context.verify_mode = ssl.CERT_NONE
                 self.ssl_context.check_hostname = False
+                self.ssl_context.verify_mode = ssl.CERT_NONE
 
                 self.logger.warning("User has enabled SSL but disabled certificate verification. This is dangerous but may be ok for a "
                                     "benchmark. Disabling urllib warnings now to avoid a logging storm. "
@@ -156,8 +155,9 @@ class OsClientFactory:
                 # advised. See: https://urllib3.readthedocs.io/en/latest/advanced-usage.html#ssl-warnings"
                 urllib3.disable_warnings()
             else:
+                # check_hostname should be False if any host is an IP address
+                self.ssl_context.check_hostname = self._has_only_hostnames(hosts)
                 self.ssl_context.verify_mode=ssl.CERT_REQUIRED
-                self.ssl_context.check_hostname = True
                 self.logger.info("SSL certificate verification: on")
 
             # When using SSL_context, all SSL related kwargs in client options get ignored
@@ -209,6 +209,20 @@ class OsClientFactory:
 
         if self._is_set(self.client_options, "enable_cleanup_closed"):
             self.client_options["enable_cleanup_closed"] = convert.to_bool(self.client_options.pop("enable_cleanup_closed"))
+
+    @staticmethod
+    def _has_only_hostnames(hosts):
+        has_ip, has_hostname = False, False
+        for host in hosts:
+            if is_ipaddress(host["host"]):
+                has_ip = True
+            else:
+                has_hostname = True
+
+        if has_ip and has_hostname:
+            raise exceptions.SystemSetupError("Could not verify certs since both IP addresses and hostnames were provided. Please ensure one or the other is used.")
+
+        return has_hostname
 
     def _is_set(self, client_opts, k):
         try:

--- a/osbenchmark/client.py
+++ b/osbenchmark/client.py
@@ -223,9 +223,9 @@ class OsClientFactory:
                 has_hostname = True
 
         if has_ip and has_hostname:
-            logger.warning("Although certificate verification is enabled, "
+            console.warn("Although certificate verification is enabled, "
                 "peer hostnames will not be matched since the host list is a mix "
-                "of names and IP addresses")
+                "of names and IP addresses", logger=logger)
             return False
 
         return has_hostname

--- a/tests/client_test.py
+++ b/tests/client_test.py
@@ -288,21 +288,7 @@ class OsClientFactoryTests(TestCase):
 
         self.assertDictEqual(original_client_options, client_options)
 
-    def test_raise_error_when_ssl_encounter_both_ip_and_hostname(self):
-        hosts = [{"host": "127.0.0.1", "port": 9200}, {"host": "localhost", "port": 9200}]
-        client_options = {
-            "use_ssl": True,
-            "verify_certs": True,
-            "http_auth": ("user", "password"),
-        }
-
-        with pytest.raises(
-            exceptions.SystemSetupError,
-            match="Can't verify certs since both IP addresses and hostnames were found.",
-        ):
-            client.OsClientFactory(hosts, client_options)
-
-    def test_check_hostname_set_to_false_when_ssl_encounter_ip(self):
+    def test_check_hostname_set_to_false_when_ssl_encounters_ips_only(self):
         hosts = [{"host": "127.0.0.1", "port": 9200}]
         client_options = {
             "use_ssl": True,
@@ -313,6 +299,19 @@ class OsClientFactoryTests(TestCase):
         f = client.OsClientFactory(hosts, client_options)
         assert f.hosts == hosts
         assert f.ssl_context.check_hostname is False
+        assert f.ssl_context.verify_mode == ssl.CERT_REQUIRED
+
+    def test_check_hostname_set_to_true_when_ssl_encounters_hostnames_only(self):
+        hosts = [{"host": "localhost", "port": 9200}]
+        client_options = {
+            "use_ssl": True,
+            "verify_certs": True,
+            "http_auth": ("user", "password"),
+        }
+
+        f = client.OsClientFactory(hosts, client_options)
+        assert f.hosts == hosts
+        assert f.ssl_context.check_hostname is True
         assert f.ssl_context.verify_mode == ssl.CERT_REQUIRED
 
 class RequestContextManagerTests(TestCase):

--- a/tests/client_test.py
+++ b/tests/client_test.py
@@ -314,6 +314,19 @@ class OsClientFactoryTests(TestCase):
         assert f.ssl_context.check_hostname is True
         assert f.ssl_context.verify_mode == ssl.CERT_REQUIRED
 
+    def test_check_hostname_set_to_false_when_ssl_encounters_both_ips_and_hostnames(self):
+        hosts = [{"host": "localhost", "port": 9200}, {"host": "127.0.0.1", "port": 9200}]
+        client_options = {
+            "use_ssl": True,
+            "verify_certs": True,
+            "http_auth": ("user", "password"),
+        }
+
+        f = client.OsClientFactory(hosts, client_options)
+        assert f.hosts == hosts
+        assert f.ssl_context.check_hostname is False
+        assert f.ssl_context.verify_mode == ssl.CERT_REQUIRED
+
 class RequestContextManagerTests(TestCase):
     @pytest.mark.skip(reason="latency is system-dependent")
     @run_async

--- a/tests/client_test.py
+++ b/tests/client_test.py
@@ -43,7 +43,7 @@ class OsClientFactoryTests(TestCase):
     cwd = os.path.dirname(__file__)
 
     def test_create_http_connection(self):
-        hosts = [{"host": "127.0.0.1", "port": 9200}]
+        hosts = [{"host": "localhost", "port": 9200}]
         client_options = {}
         # make a copy, so we can verify later that the factory did not modify it
         original_client_options = dict(client_options)
@@ -59,7 +59,7 @@ class OsClientFactoryTests(TestCase):
 
     @mock.patch.object(ssl.SSLContext, "load_cert_chain")
     def test_create_https_connection_verify_server(self, mocked_load_cert_chain):
-        hosts = [{"host": "127.0.0.1", "port": 9200}]
+        hosts = [{"host": "localhost", "port": 9200}]
         client_options = {
             "use_ssl": True,
             "verify_certs": True,
@@ -94,7 +94,7 @@ class OsClientFactoryTests(TestCase):
 
     @mock.patch.object(ssl.SSLContext, "load_cert_chain")
     def test_create_https_connection_verify_self_signed_server_and_client_certificate(self, mocked_load_cert_chain):
-        hosts = [{"host": "127.0.0.1", "port": 9200}]
+        hosts = [{"host": "localhost", "port": 9200}]
         client_options = {
             "use_ssl": True,
             "verify_certs": True,
@@ -136,7 +136,7 @@ class OsClientFactoryTests(TestCase):
 
     @mock.patch.object(ssl.SSLContext, "load_cert_chain")
     def test_create_https_connection_only_verify_self_signed_server_certificate(self, mocked_load_cert_chain):
-        hosts = [{"host": "127.0.0.1", "port": 9200}]
+        hosts = [{"host": "localhost", "port": 9200}]
         client_options = {
             "use_ssl": True,
             "verify_certs": True,
@@ -170,7 +170,7 @@ class OsClientFactoryTests(TestCase):
         self.assertDictEqual(original_client_options, client_options)
 
     def test_raises_error_when_only_one_of_client_cert_and_client_key_defined(self):
-        hosts = [{"host": "127.0.0.1", "port": 9200}]
+        hosts = [{"host": "localhost", "port": 9200}]
         client_options = {
             "use_ssl": True,
             "verify_certs": True,
@@ -211,7 +211,7 @@ class OsClientFactoryTests(TestCase):
 
     @mock.patch.object(ssl.SSLContext, "load_cert_chain")
     def test_create_https_connection_unverified_certificate(self, mocked_load_cert_chain):
-        hosts = [{"host": "127.0.0.1", "port": 9200}]
+        hosts = [{"host": "localhost", "port": 9200}]
         client_options = {
             "use_ssl": True,
             "verify_certs": False,
@@ -248,7 +248,7 @@ class OsClientFactoryTests(TestCase):
 
     @mock.patch.object(ssl.SSLContext, "load_cert_chain")
     def test_create_https_connection_unverified_certificate_present_client_certificates(self, mocked_load_cert_chain):
-        hosts = [{"host": "127.0.0.1", "port": 9200}]
+        hosts = [{"host": "localhost", "port": 9200}]
         client_options = {
             "use_ssl": True,
             "verify_certs": False,
@@ -288,6 +288,32 @@ class OsClientFactoryTests(TestCase):
 
         self.assertDictEqual(original_client_options, client_options)
 
+    def test_raise_error_when_ssl_encounter_both_ip_and_hostname(self):
+        hosts = [{"host": "127.0.0.1", "port": 9200}, {"host": "localhost", "port": 9200}]
+        client_options = {
+            "use_ssl": True,
+            "verify_certs": True,
+            "http_auth": ("user", "password"),
+        }
+
+        with pytest.raises(
+            exceptions.SystemSetupError,
+            match="Could not verify certs since both IP addresses and hostnames were provided. Please ensure one or the other is used.",
+        ):
+            client.OsClientFactory(hosts, client_options)
+
+    def test_check_hostname_set_to_false_when_ssl_encounter_ip(self):
+        hosts = [{"host": "127.0.0.1", "port": 9200}]
+        client_options = {
+            "use_ssl": True,
+            "verify_certs": True,
+            "http_auth": ("user", "password"),
+        }
+
+        f = client.OsClientFactory(hosts, client_options)
+        assert f.hosts == hosts
+        assert f.ssl_context.check_hostname is False
+        assert f.ssl_context.verify_mode == ssl.CERT_REQUIRED
 
 class RequestContextManagerTests(TestCase):
     @pytest.mark.skip(reason="latency is system-dependent")

--- a/tests/client_test.py
+++ b/tests/client_test.py
@@ -298,7 +298,7 @@ class OsClientFactoryTests(TestCase):
 
         with pytest.raises(
             exceptions.SystemSetupError,
-            match="Could not verify certs since both IP addresses and hostnames were provided. Please ensure one or the other is used.",
+            match="Can't verify certs since both IP addresses and hostnames were found.",
         ):
             client.OsClientFactory(hosts, client_options)
 

